### PR TITLE
Allow the issuer to select an SMS template

### PR DESCRIPTION
**Pull request by whaught**
Issue https://github.com/google/exposure-notifications-verification-server/issues/1320

## Proposed Changes

* If more than one SMS template is defined, allow the user to pick one
* Add a field to the issue API and compose with the given template

**Release Note**

```release-note
Allow user to select an SMS template on code-issue. Add template label field to issueAPI.
```

### CODE REVIEW

File: cmd/server/assets/codes/issue.html [hunk 1]
```diff
@@ -123,8 +123,26 @@ <h1>{{t $.locale "codes.issue.header"}}</h1>
...
+            {{if $currentRealm.SMSTextAlternateTemplates}}
             <div class="row form-group">
-              <label for="symptomDate" class="col-sm-6 col-md-4 col-lg-3">{{t $.locale "codes.issue.sms-text-message-label"}}</label>
+              <label for="sms-template" class="col-sm-6 col-md-4 col-lg-3">{{t $.locale "codes.issue.sms-template-label"}}</label>
+              <div class="col-sm-6 col-md-8 col-lg-9">
+                <div class="input-group">
+                  <select class="form-control" name="sms-template">
+                    <option value="Default SMS template">Default SMS template</option>
+                    {{range $k, $v := $currentRealm.SMSTextAlternateTemplates}}
+                    <option value="{{$k}}">{{$k}}</option>
```
Comment (sethvargo) [inline cmd/server/assets/codes/issue.html:134]: Nit: indent

Comment (sethvargo) [inline cmd/server/assets/codes/issue.html:134]: Should this be `{{$v}}`? If not, why is `$v` scoped?

File: pkg/api/api.go [hunk 2]
```diff
@@ -186,8 +186,9 @@ type IssueCodeRequest struct {
 	TestType    string `json:"testType"`
 	// Offset in minutes of the user's timezone. Positive, negative, 0, or omitted
 	// (using the default of 0) are all valid. 0 is considered to be UTC.
-	TZOffset float32 `json:"tzOffset"`
-	Phone    string  `json:"phone"`
+	TZOffset         float32 `json:"tzOffset"`
+	Phone            string  `json:"phone"`
+	SMSTemplateLabel string  `json:"SMSTemplateLabel"`
```
Comment (sethvargo) [inline pkg/api/api.go:191]: This should be `smsTemplateLabel` (I don't make the rules)

File: pkg/controller/issueapi/logic.go [hunk 3]
```diff
@@ -260,7 +260,11 @@ func (c *Controller) issue(ctx context.Context, authApp *database.AuthorizedApp,
 		if err := func() error {
 			defer observability.RecordLatency(ctx, time.Now(), mSMSLatencyMs, &result.obsBlame, &result.obsResult)
 
-			message := realm.BuildSMSText(code, longCode, c.config.GetENXRedirectDomain())
+			message, err := realm.BuildSMSText(code, longCode, c.config.GetENXRedirectDomain(), request.SMSTemplateLabel)
+			if err != nil {
+				logger.Warnw("failed to compose SMS", "error", err)
```
Comment (sethvargo) [inline pkg/controller/issueapi/logic.go:265]: Just return err here, it's logged later

File: pkg/database/realm_test.go [hunk 4]
```diff
@@ -34,14 +34,14 @@ func TestSMS(t *testing.T) {
 	realm.SMSTextTemplate = "This is your Exposure Notifications Verification code: [enslink] Expires in [longexpires] hours"
 	realm.RegionCode = "US-WA"
 
-	got := realm.BuildSMSText("12345678", "abcdefgh12345678", "en.express")
+	got, _ := realm.BuildSMSText("12345678", "abcdefgh12345678", "en.express", "")
```
Comment (sethvargo) [inline pkg/database/realm_test.go:37]: Check error here and below

File: internal/i18n/locales/en/default.po [hunk 5]
```diff
@@ -102,6 +102,12 @@ msgstr "SMS text message (recommended)"
 msgid "codes.issue.sms-text-message-label"
 msgstr "Patient phone number"
 
+msgid "codes.issue.sms-template-label"
```
Comment (sethvargo) [inline internal/i18n/locales/en/default.po:105]: I _think_ you need to translate all the things now or the tests will fail.

**Review by sethvargo — CHANGES_REQUESTED** [review]

File: cmd/server/assets/codes/issue.html (hunk 1 above)
Comment (whaught) [reply cmd/server/assets/codes/issue.html:134]: templates can't do `$k, _` and single var is just values. I need just the keys (labels)

File: pkg/api/api.go (hunk 2 above)
Comment (whaught) [reply pkg/api/api.go:191]: yea that's consistent with tzOffset. Weird though.

File: cmd/server/assets/codes/issue.html (hunk 1 above)
Comment (sethvargo) [reply cmd/server/assets/codes/issue.html:134]: but then can do `$k := range`

Comment (whaught) [reply cmd/server/assets/codes/issue.html:134]: but `$k := range .someMap` iterates the map values, not the keys. Maybe I'm misunderstanding

File: internal/i18n/locales/de/default.po [hunk 6]
```diff
@@ -102,6 +102,12 @@ msgstr "SMS Textnachricht (empfohlen)"
 msgid "codes.issue.sms-text-message-label"
 msgstr "Patienten Telefonnummer"
 
+msgid "codes.issue.sms-template-label"
+msgstr "SMS Vorlage"
+
+msgid "codes.issue.sms-template-detail"
+msgstr "Der Patient erhält eine SMS mit der ausgewählten Nachrichtenvorlage."
```
Comment (whaught) [inline internal/i18n/locales/de/default.po:109]: I don't know enough german to refute that `ausgewählten Nachrichtenvorlage` is `selected template` but it sure looks long

File: internal/i18n/locales/en/default.po (hunk 5 above)
Comment (whaught) [reply internal/i18n/locales/en/default.po:105]: Added Google translate translations.

File: internal/i18n/locales/de/default.po (hunk 6 above)
Comment (sethvargo) [reply internal/i18n/locales/de/default.po:109]: @hostirosti could you sanity check these two new German translations please?

Comment (hostirosti) [reply internal/i18n/locales/de/default.po:109]: LGTM

File: cmd/server/assets/codes/issue.html (hunk 1 above)
Comment (sethvargo) [reply cmd/server/assets/codes/issue.html:134]: My b - apparently this is backwards in Go templates from the Go language /tableflip

File: docs/api.md [hunk 7]
```diff
@@ -207,6 +208,10 @@ Request a verification code to be issued. Accepts [optional] symptom date and te
 * `phone`
   * Phone number to send the SMS to. If a phone number is provided, but the SMS text
     message fails to send, the API will return a 4xx client error.
+* `smsTemplateLabel`
+  * If the realm has more than one SMS template defined, this may be optionally specify
+    the label of the message template which the server should compose. If omitted, the
+    default template will be used.
```
Comment (sethvargo) [inline docs/api.md:214]: What happens if they supply a label that does not exist?

**Review by sethvargo — APPROVED** [review]
/lgtm

@whaught can do you do a followup to update the docs on the failure modes?

**Pull request merged**